### ISSUE TITLE
AccessVioltationException while parsing file

### ISSUE THREAD

**Issue by zillemarco** [mono/CppSharp#959]
The C++ parse throws a `System.AccessViolationException` when parsing files.
The error is sent from `Parser.cs@66: var result = Parser.ClangParser.ParseHeader(options);`.

It doesn't always happen, for small files it's ok but as soon as I try to parse a (not too much) complex file, which includes a lot of other files, this error is thrown.

One thing that comes to my mind: could it be that since I'm using a 32bit compilation and the file includes a lot of other headers I run out of memory? For a single file, task manager shows me almost 200MB of memory used.

**Comment by ddobrev** [comment]
@zillemarco two items come to mind:
1. Upload your project or an equivalent so that we can take a look;
2. If you use more than one header, try the option for a unity build.

**Comment by zillemarco** [comment]
@ddobrev I've tried with unity build but I have problems with clang which gives me errors saying that some my classes, enum and structs are redefined, even though I use `#pragma once` on every file.

About uploading the project, well, it's quite big and unfortunately is owned by the company I work for, so I can't upload it unfortunately. Do you have some hints on what I could try to debug? Otherwise I'll try to reproduce the same error on another project and upload it.

In the meantime I'm trying to compile a 64bit CppSharp to see if it works.

**Comment by ddobrev** [comment]
@zillemarco don't waste your time, the NuGet already is 64-bit. Or do you mean you cannot use it because of the bug you fixed yesterday?
About the unity build and #pragma once: these redefinitions are strange. Unless it's too much work, could you temporarily change them to header guards instead and try again?

**Comment by zillemarco** [comment]
@ddobrev yeah, I can't use the NuGet because of the bug and yes, I was just looking that on Windows it is defaulted to build at 32bit :D

About the `#pragma once` I can't change them to header guard in the project since it would be a pain but I guess I can set up something similar without problems.

**Comment by ddobrev** [comment]
@zillemarco I'd rather not release a new NuGet yet because I prefer to collect a larger number of fixes. However, I can upload you a 64-bit archive, it's easy to me. Would you like to?

**Comment by zillemarco** [comment]
@ddobrev yes, I supposed a new NuGet wasn't doable :)

If you could give me a 64bit archive it is very welcome, thank you :D

**Comment by ddobrev** [comment]
There you go:

[CppSharp.zip](https://github.com/mono/CppSharp/files/1320343/CppSharp.zip)

This is for your fix yesterday. :)

**Comment by zillemarco** [comment]
Thanks :D I'll let you know if this works

**Comment by zillemarco** [comment]
@ddobrev that's it. With the 64bit build it doesn't crash. I have other problems with errors clang gives me, but it doesn't crash and there is no sign of `AccessViolationException`

**Comment by ddobrev** [comment]
@zillemarco I am happy it's worked.
It makes some sense because we do use a substantial amount of memory, we can't get to optimising it. Perhaps the 32-bit C++# cannot address enough.
Have you had any other trouble so far? I hope you don't need to manually fix the generated code.

**Comment by zillemarco** [comment]
@ddobrev actually, I just fixed the errors clang was giving me and I tried again to parse that single file. I don't get any exception, the program crashes always after it reaches about 325MB (I see that from Visual Studio diagnostics running in debug). I am certain that it is still parsing since both the `Postprocess` and the `Preprocess` method of my `ILibrary` aren't yet called when it crashes, but I don't know the exact point because the program just closes itself, no errors, no messages.

And I was so happy :(

BTW: this happens even with unity build turned on.

**Comment by ddobrev** [comment]
@zillemarco I don't quite understand - are you saying it crashes with the 64-bit version as well?

**Comment by zillemarco** [comment]
@ddobrev Yes exactly. Could it be that I don't get any error from Visual Studio since I'm missing the `.pdb`?

**Comment by zillemarco** [comment]
@ddobrev the documentation doesn't explain this but, looking at the scripts to download the dependencies, to create a 64bit build of CppSharp do I just need to modify `DownloadDeps.bat` to pass `amd64`? It that's so I can compile the 64bit version myself and try to address this problem.

**Comment by ddobrev** [comment]
@zillemarco what's the point when you get the same problem with 64-bit? You might as well try the 32-bit build you already have.

**Comment by ddobrev** [comment]
These are the PDB-s for 64-bit, I hope they can help: https://www.filedog.io/NiceFlea (too large to upload here).

**Comment by zillemarco** [comment]
@ddobrev thanks for the PDB-s. I was able to assert that the problem is still inside the CppParser ParseHeaders method but I can't step into it while debugging. I'll try debugging the 32bit version I have. One question: even when using CppSharp as the running project (launching my program to debug it) I still can't step into the ParseHeader function. Do you set the `CppSharp.CppParser` native project as startup project to debug it?

**Comment by ddobrev** [comment]
@zillemarco CppSharp.CppParser is a library, you cannot set it as the starting project. I am not sure why you cannot step in because I have uploaded the PDB of the native library as well.

**Comment by zillemarco** [comment]
@ddobrev I can set it as the startup project if I change the debug command to launch my program instead of the library itself. BTW the problem isn't with your PDB-s, even from my debug build I couldn't stop into that function.

**Comment by ddobrev** [comment]
Have you marked "Enable native debugging" in your startup project?

**Comment by zillemarco** [comment]
Thanks @ddobrev that did the trick. I was able to track the error down to 
`ASTContext.h@1914: TypeInfo getTypeInfo(QualType T) const { return getTypeInfo(T.getTypePtr()); }` but no more because I don't have the clang sources.

Basically a call to `clang::ASTContext::getTypeInfoImpl` fails on this macro `clang/include/clang/AST/TypeNodes.def@89: NON_CANONICAL_UNLESS_DEPENDENT_TYPE(UnaryTransform, Type)`, whick is resolved to `clang::ASTContext::getASTRecordLayout`

Sorry I can't tell you what happens inside `clang::ASTContext::getASTRecordLayout` but, as I said, I don't have the clang sources :(

However, I am certain this is happening while walking a templated class specialization. I have tried setting `GenerateClassTemplates` to `false`, but nothing changes.

**Comment by ddobrev** [comment]
@zillemarco I would guess there's a type of type (yeah, that's right!) we don't support yet. If you can figure it out, add support for it. If not, what you could do is start reducing your test case i.e. your header while keeping the bug reproducible. You'd end up with a really small sample which you could share.

**Comment by zillemarco** [comment]
@ddobrev mmmm...do you mean a templated class that contains a specialization of another templated class? If that's what you mean it's simpler for me to try to reproduce it without touching my project (it's full of these unfortunately).

**Comment by ddobrev** [comment]
@zillemarco I am not sure and I really wouldn't try figuring it out without a test case, not for free anyway. If you cannot either, that's fine, just go with the second approach. When you have isolated a test case, I'll help you.

**Comment by zillemarco** [comment]
@ddobrev no problems. I'll prepare a test case :)

**Comment by zillemarco** [comment]
@ddobrev I have a test case! :D

Not an extremely simple one but I found a piece of one of the classes inside my project which had the problem and I've extrapolated it.

I was then able to strip it further down until I got something that CppSharp was able to handle.

How do you want to operate?

**Comment by tritao** [comment]
Can you post the code in a gist or something?

**Comment by zillemarco** [comment]
Yes sure. I have prepared 5 files, from the start (the most complex) down to the last one which compiles. Do you want all of them or only the non-working simplest and the working one?

**Comment by ddobrev** [comment]
@zillemarco the simplest reproduction should do.

**Comment by zillemarco** [comment]
Here we go.

This is the simplest not working: [here](https://gist.github.com/zillemarco/c6e60f69083f1a2978fe0db0a7a40342)

This is the fixed (just for comparison): [here](https://gist.github.com/zillemarco/7529e9b5657fb801c8b385233bf9eadf)

**Comment by tritao** [comment]
I can repro on OSX:

```
Native stacktrace:

	0   mono                                0x00000001092a5c91 mono_handle_native_crash + 257
	1   mono                                0x000000010930baf6 altstack_handle_and_restore + 70
	2   libCppSharp.CppParser.dylib         0x000000010df728e0 _ZN12_GLOBAL__N_126ItaniumRecordLayoutBuilder16InitializeLayoutEPKN5clang4DeclE + 16
	3   libCppSharp.CppParser.dylib         0x000000010df6bd0a _ZNK5clang10ASTContext18getASTRecordLayoutEPKNS_10RecordDeclE + 3882
	4   libCppSharp.CppParser.dylib         0x000000010ddc7eab _ZNK5clang10ASTContext15getTypeInfoImplEPKNS_4TypeE + 1403
	5   libCppSharp.CppParser.dylib         0x000000010ddc8a2e _ZNK5clang10ASTContext11getTypeInfoEPKNS_4TypeE + 174
	6   libCppSharp.CppParser.dylib         0x000000010ddc7e21 _ZNK5clang10ASTContext15getTypeInfoImplEPKNS_4TypeE + 1265
	7   libCppSharp.CppParser.dylib         0x000000010ddc8a2e _ZNK5clang10ASTContext11getTypeInfoEPKNS_4TypeE + 174
	8   libCppSharp.CppParser.dylib         0x000000010d1ad342 _ZN8CppSharp9CppParser6Parser8WalkTypeEN5clang8QualTypeEPKNS2_7TypeLocEb + 1218
	9   libCppSharp.CppParser.dylib         0x000000010d1b5002 _ZN8CppSharp9CppParser6Parser12WalkVariableEPKN5clang7VarDeclEPNS0_3AST8VariableE + 322
	10  libCppSharp.CppParser.dylib         0x000000010d1b4a72 _ZN8CppSharp9CppParser6Parser12WalkVariableEPKN5clang7VarDeclE + 130
	11  libCppSharp.CppParser.dylib         0x000000010d1aaeb9 _ZN8CppSharp9CppParser6Parser15WalkDeclarationEPKN5clang4DeclEbb + 985
	12  libCppSharp.CppParser.dylib         0x000000010d1b0975 _ZN8CppSharp9CppParser6Parser10WalkRecordEPKN5clang10RecordDeclEPNS0_3AST5ClassE + 501
	13  libCppSharp.CppParser.dylib         0x000000010d1b023a _ZN8CppSharp9CppParser6Parser13WalkRecordCXXEPKN5clang13CXXRecordDeclEPNS0_3AST5ClassE + 74
	14  libCppSharp.CppParser.dylib         0x000000010d1aad6d _ZN8CppSharp9CppParser6Parser15WalkDeclarationEPKN5clang4DeclEbb + 653
	15  libCppSharp.CppParser.dylib         0x000000010d1b8b45 _ZN8CppSharp9CppParser6Parser11ParseHeaderERKNSt3__16vectorINS2_12basic_stringIcNS2_11char_traitsIcEENS2_9allocatorIcEEEENS7_IS9_EEEEPNS0_12ParserResultE + 1637
```

**Comment by ddobrev** [comment]
@zillemarco does AlignedBytes affect anything at all?

**Comment by zillemarco** [comment]
@ddobrev what do you mean? If it is used somewhere?

**Comment by ddobrev** [comment]
@zillemarco I saw where it was used, I mean if the bug occurs if it's any simpler.

**Comment by tritao** [comment]
Smaller test case:

```
template<typename OptionalType>
struct Optional
{
};

class Test
{
private:
    static Optional<int> s_touchIDs[10];
};
```

**Comment by zillemarco** [comment]
@ddobrev well....that's @tritao for you :D

**Comment by zillemarco** [comment]
@ddobrev yes, no problems. I'll add @tritao 's smaller test case at the end of `Common.h` and make a pull request if that's ok.

Debugging earlier I've more or less located the point where the problem could be but I couldn't go any further since I have to compile clang on my PC. I'll do some other tests as soon as I compile clang.

BTW, the problem seems to be that it can't compute the size of the element of the array, the size of `struct Optional` in this case.

**Comment by ddobrev** [comment]
@zillemarco another important clue: it depends on the Clang version. I use a little older one for my 64-bit build because of unrelated reasons. Mine is LLVM 616d6a1621f160f27847e0332306ccff2ee43ac7 and Clang 158a1fe17bb242d8c0555c81de2eb226ad62c9da. The ones in our packages, and which you use, are written in the LLVM-commit and Clang-commit files.

**Comment by ddobrev** [comment]
@zillemarco @tritao and I have found the reason. We need to force completion of template specialisations. We need this feature anyway because of our support for templates. I'll try to have it done by the end of the week because there are some complications.

**Comment by tritao** [comment]
Until @ddobrev is done, there is a workaround you can try in `Parser.cpp`, replace the existing method with this version:

```
bool Parser::ShouldCompleteType(const clang::QualType& QualType, bool LocValid)
{
    auto FinalType = GetFinalType(QualType.getTypePtr());
    if (auto Tag = FinalType->getAsTagDecl())
    {
        if (auto CTS = llvm::dyn_cast<clang::ClassTemplateSpecializationDecl>(Tag))
        {
            // we cannot get a location in some cases of template arguments
            if (!LocValid || CTS->isCompleteDefinition())
                return false;

            auto TAL = &CTS->getTemplateArgs();
            for (size_t i = 0; i < TAL->size(); i++)
            {
                auto TA = TAL->get(i);
                if (TA.getKind() == clang::TemplateArgument::ArgKind::Type)
                {
                    auto Type = TA.getAsType();
                    if (Type->isVoidType())
                        return false;
                }
            }
            auto Unit = GetTranslationUnit(Tag);
            // HACK: completing all system types overflows the managed stack
            // while running the AST converter since the latter is a giant indirect recursion
            // this solution is a hack because we might need to complete system template specialisations
            // such as std:string or std::vector in order to represent them in the target language
            return !Unit->isSystemHeader;
        }
    }
    return false;
}
```

**Comment by zillemarco** [comment]
@tritao I've tried the change you have suggested but even the CppSharp tests themselves give me `AccessViolationException`. Don't know if it could be the problem with 32bit build running out of memory.

**Comment by ddobrev** [comment]
@zillemarco what happens if you try it with just the test case?

**Comment by zillemarco** [comment]
@ddobrev just tried for that test case only and it works. The other test cases inside CppSharp give `AccessViolationException` like the ones I got yesterday, but with the code @tritao suggested, my specific test case works fine even on 32bit build.

**Comment by tritao** [comment]
Yeah the code still needs some work as a general solution, @ddobrev is still working it out, but it might be enough for your specific issue for now.

**Comment by zillemarco** [comment]
@tritao Yeah of course. I'll give it a try to see if it works for me for now.

**Comment by ddobrev** [comment]
@zillemarco @tritao 's right. I would've replied earlier but got distracted with some work. So, just forget about our tests. As I said last night and as @tritao 's just reminded, the complete fix needs more work, otherwise we would've just pushed it instead of pasting you code. So as long as it works with your real header, we're all fine for the time being. Does it?

**Comment by zillemarco** [comment]
@ddobrev As soon as I get home I'll try because seeing the tests not working I didn't try with my complete code. I'll let you know

**Comment by zillemarco** [comment]
@ddobrev unfortunately it still doesn't work on my project. It has troubles handling a template specialization for `std::is_arithmetic`. If I can create a new test case to address this one too I'll report it.

**Comment by ddobrev** [comment]
@zillemarco please do, thank you.

**Comment by zillemarco** [comment]
@ddobrev no problems at all. Those damn templates are quite difficult to handle :D

**Comment by ddobrev** [comment]
@zillemarco you cannot imagine...

**Comment by ddobrev** [comment]
@zillemarco any luck isolating that second problem?

**Comment by zillemarco** [comment]
@ddobrev to be honest I didn't have time to do it this weekend. Now I'm about to finish at work and I'll try.

**Comment by ddobrev** [comment]
@zillemarco I thought all of this was related to your job. Anyway, thank you for your effort, it would be great to fix the second bug as well.

**Comment by zillemarco** [comment]
@ddobrev well, it is related to my job but saturday they gave me an urgent task (damn...) so I'm working on it at home

**Comment by zillemarco** [comment]
@ddobrev found it. I noticed the second problem appeared even only including type_trais from the standard library (`#include <type_traits>` is enough). However I restricted the error finding out is has something to do with derivate a template class from a templated class specializing one of the arguments, etc....I don't know how to explain it :D

Here are gists showing a not working and two working examples, maybe they are of help...I don't really know how to explain this behaviour:

- not working: [here](https://gist.github.com/zillemarco/e8eb59ff217c27420a114ddea348bfe7)
- first fix: [here](https://gist.github.com/zillemarco/8d9b203eca79402a588b7064d840b3b1)
- second fix (this seems more interesting to me): [here](https://gist.github.com/zillemarco/e3a498d586df342533ebb6a49462ad30)

**Comment by ddobrev** [comment]
@zillemarco thank you very much. May I further ask of you to report a separate issue about this problem, along with the symptom and the test case?

**Comment by zillemarco** [comment]
@ddobrev yes of course.
